packages/client/src/helpers/runWithRetry.ts

```diff
@@ -0,0 +1,181 @@
+ * holds error data, if it does, it throws `RpcError` to prevent false positives for `runWithRetry` retry handler.
+ */
+export const handleFalsePositiveResponse = <
+  I extends object,
+  O extends SfuResponseWithError,
+  T extends FunctionToRetry<FinishedUnaryCall<I, O>>,
+>(
+  f: T,
+) =>
+  async function falsePositiveHandler(...functionArguments: Parameters<T>) {
+    // await or throw if "f" fails
+    const data = (await f(...functionArguments)) as Awaited<ReturnType<T>>;
+    // check for error data, throw if exist
+    if (data.response.error) {
+      throw new RpcError(
+        data.response.error.message,
+        TwirpErrorCode.unknown.toString(),
+        // @ts-ignore - RpcError only allows for values to be string (it parses these metadata for logging)
+        // though boolean parses to string so it's fine to ignore
+        { shouldRetry: data.response.error.shouldRetry },
+      );
+    }
+
+    return data;
+  };
+
+export type RunWithRetryOptions<T extends FunctionToRetry> = {
+  retryAttempts?: number;
+  delayBetweenRetries?: number | ((attempt: number) => number);
+  isRetryable?: (error: unknown) => boolean;
+  didValueChange?: (
+    ...functionArguments: Parameters<T>
+  ) => Promise<boolean> | boolean;
+};
+
+/**
+ * Function which wraps asynchronous functions with retry mechanism which'll keep executing said
+ * function pre-defined number of times until it resolves, rejects or the retry mechanism runs out of available attempts.
+ *
+ * #### Available options:
+ *  - `didValueChange` - check function with initial argument value to check against new values, return `true` to abort next attempt
+ *
+ *  - `isRetryable` - error evaluation function which determines whether to retry the execution
+ *
+ *  - `delayBetweenRetries` - accepts either fixed numeric value or function which provides retry attempt as the argument, expects number as return value
+ *
+ *  - `retryAttempts` - number of attempts to try out before rejecting the promise
+ */
+export const runWithRetry = <T extends FunctionToRetry>(
+  f: T,
+  {
+    retryAttempts = 3,
+    delayBetweenRetries,
+    isRetryable,
+    didValueChange,
+  }: RunWithRetryOptions<T> = {},
+) =>
+  async function retryable(...functionArguments: Parameters<T>) {
+    // starting with -1 as first attempt is not considered a retry
+    let retryAttempt = -1;
+
+    do {
+      let data: Awaited<ReturnType<T>> | null = null;
+      let error: unknown = null;
+      const isLastRun = retryAttempts === retryAttempt + 1;
+      const isInitialRun = retryAttempt === -1;
+
+      if (await didValueChange?.(...functionArguments)) {
+        throw new RetryError({ type: 'abort' });
+      }
+
+      try {
+        data = await f(...functionArguments);
+      } catch (e) {
+        error = e;
+      }
+
+      if (await didValueChange?.(...functionArguments)) {
+        throw new RetryError({ type: 'abort' });
+      }
```

Comment:
I've been thinking and realized that this abort check might be misleading and would cause stale-ish client-side state if the call goes through; local state would not represent server sate.